wildmatch.go

```diff
@@ -141,7 +153,43 @@ func escapable(c byte) bool {
 
 // parseTokens parses a separated list of patterns into a sequence of
 // representative Tokens that will compose the pattern when applied in sequence.
-func parseTokens(dirs []string) []token {
+func (w *Wildmatch) parseTokens(dirs []string) []token {
+	if len(dirs) == 0 {
+		return make([]token, 0)
+	}
+
+	endRange := len(dirs)
+	var finalComponents []token
+
+	// This is a pattern like "foo/" or "foo/bar/".
+	if !w.gitattributes && len(dirs) > 1 && dirs[len(dirs)-1] == "" {
+		if w.contents {
+			finalComponents = []token{&trailingComponents{}}
+		}
+		if len(dirs) == 2 {
+			// We don't have a slash in the middle, so this can go
+			// anywhere in the hierarchy.  If there had been a slash
+			// here, it would have been anchored at the root.
+			rest := w.parseTokensSimple(dirs)
+			tokens := append([]token{&unanchoredDirectory{
+				Until: rest[0],
+			}})
+			// If we're not matching all contents, then do include
+			// the empty component so we don't match
+			// non-directories.
+			if finalComponents == nil && len(rest) > 1 {
+				finalComponents = rest[1:]
+			}
+			tokens = append(tokens, finalComponents...)
+			return tokens
+		}
+	}
+	components := w.parseTokensSimple(dirs[0:endRange])
+	components = append(components, finalComponents...)
+	return components
```

Comment:
I believe `endRange` can be dropped as it's only used once and `dirs` will suffice in that spot, and I wondered if it would be possible to take advantage of the similarity of the `tokens = append(tokens, finalComponents...)` and `components = append(components, finalComponents...)` lines and avoid the early `return tokens`.  I experimented a bit and came up with the following; see if this makes sense to you, though, and doesn't introduce any new errors!
```suggestion
	components := w.parseTokensSimple(dirs)
	var finalComponents []token

	// This is a pattern like "foo/" or "foo/bar/".
	if !w.gitattributes && len(dirs) > 1 && dirs[len(dirs)-1] == "" {
		if w.contents {
			finalComponents = []token{&trailingComponents{}}
		}
		if len(dirs) == 2 {
			// If we're not matching all contents, then do include
			// the empty component so we don't match
			// non-directories.
			if !w.contents && len(components) > 1 {
				finalComponents = components[1:]
			}
			// We don't have a slash in the middle, so this can go
			// anywhere in the hierarchy.  If there had been a slash
			// here, it would have been anchored at the root.
			components = append([]token{&unanchoredDirectory{
				Until: components[0],
			}})
		}
	}

	return append(components, finalComponents...)
```

Reply:
Fixed.  I originally thought that would be necessary to make things work, but found it was not.